flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java

```diff
@@ -0,0 +1,235 @@
+     * logic in {@code submitRequestsToApi}.
+     *
+     * <p>There is a limit on the number of concurrent (async) requests that can be handled by the
+     * client library. This limit is enforced by checking the queue size before accepting a new
+     * element into the queue.
+     *
+     * <p>To complete a checkpoint, we need to make sure that no requests are in flight, as they may
+     * fail, which could then lead to data loss.
+     */
+    private int inFlightRequestsCount;
+
+    @Override
+    public void write(InputT element, Context context) throws IOException, InterruptedException {
+        while (bufferedRequestEntries.size() >= maxBufferedRequests) {
+            mailboxExecutor.yield();
+        }
+
+        bufferedRequestEntries.add(elementConverter.apply(element, context));
+
+        flushIfFull();
+    }
+
+    private void flushIfFull() throws InterruptedException, IOException {
+        while (bufferedRequestEntries.size() >= maxBatchSize) {
+            flush();
+        }
+    }
+
+    /**
+     * Persists buffered RequestsEntries into the destination by invoking {@code
+     * submitRequestEntries} with batches according to the user specified buffering hints.
+     *
+     * <p>The method blocks if too many async requests are in flight.
+     */
+    private void flush() throws InterruptedException, IOException {
+        while (inFlightRequestsCount >= maxInFlightRequests) {
+            mailboxExecutor.yield();
+        }
+
+        List<RequestEntryT> batch = new ArrayList<>(maxBatchSize);
+
+        while (batch.size() <= maxBatchSize && !bufferedRequestEntries.isEmpty()) {
+            try {
+                batch.add(bufferedRequestEntries.remove());
+            } catch (NoSuchElementException e) {
+                break;
+            }
+        }
+
+        if (batch.size() == 0) {
+            return;
+        }
+
+        ResultFuture<RequestEntryT> requestResult =
+                failedRequestEntries ->
+                        mailboxExecutor.execute(
+                                () -> completeRequest(failedRequestEntries),
+                                "Mark in-flight request as completed and requeue %d request entries",
+                                failedRequestEntries.size());
+
+        inFlightRequestsCount++;
+        try {
+            submitRequestEntries(batch, requestResult);
+        } catch (RuntimeException e) {
+            // if a runtime exception is thrown, completeRequest will not have been called
+            inFlightRequestsCount--;
+            throw new IOException(
+                    String.format(
+                            "Failed to submit up to [%s] request entries, "
+                                    + "POSSIBLE DATA LOSS. A runtime exception occured during the submission of the"
+                                    + " request entries",
+                            batch.size()),
+                    e);
```

Comment:
I was on the fence here about whether to retry that batch when a `RuntimeException` is thrown in `submitRequestEntries`. I was worried that it will just keep failing for different batches with the same broken element. So I decided to cancel the job here.

I agree I should create a dedicated `Exception` though.